src/lib/lspace/lspace_area.rs

```diff
@@ -3,274 +3,232 @@
     }
 
-    fn on_key_release(&mut self, event_key: &EventKey) {
+    pub fn on_leave(&mut self, mod_state: InputModifierState, pos: Point2) {
+        self.input_mods = mod_state;
+        self.input_pointer.set_position(PointerPosition::out_of_bounds());
     }
 
-    fn on_enter(&mut self, event_crossing: &EventCrossing) {
+    pub fn on_motion(&mut self, mod_state: InputModifierState, pos: Point2) {
+        self.input_mods = mod_state;
+        self.input_pointer.set_position(PointerPosition::at_position(pos));
     }
 
-    fn on_leave(&mut self, event_crossing: &EventCrossing) {
+    pub fn on_scroll(&mut self, mod_state: InputModifierState, pos: Point2,
+                 scroll_x: f64, scroll_y: f64) {
+        self.input_mods = mod_state;
+        self.input_pointer.set_position(PointerPosition::at_position(pos));
     }
 
-    fn on_motion(&mut self, event_motion: &EventMotion) {
+    pub fn on_key_press(&mut self, mod_state: InputModifierState, key_val: u32, key_string: String) {
+        self.input_mods = mod_state;
+        self.input_keyboard.on_key_press(mod_state, key_val, key_string);
     }
 
-    fn on_scroll(&mut self, event_scroll: &EventScroll) {
+    pub fn on_key_release(&mut self, mod_state: InputModifierState, key_val: u32, key_string: String) {
+        self.input_mods = mod_state;
+        self.input_keyboard.on_key_release(mod_state, key_val, key_string);
     }
 
-    fn on_draw(&mut self, cairo_ctx: Context) {
-        self.initialise(&cairo_ctx);
-        self.layout();
-        self.draw(&cairo_ctx);
+    pub fn on_draw(&mut self, cairo_ctx: &Context) {
+
+        self.layout(cairo_ctx);
+        self.draw(cairo_ctx);
     }
 
-    fn layout(&mut self) {
-        if self.layout_required {
-            match &self.elem {
-                &Some(ref re) => {
-                    let e = re.as_root_element().unwrap();
-                    let t1 = time::precise_time_ns();
-                    let rx = e.root_requisition_x();
-                    e.root_allocate_x(self.width as f64);
-                    let ry = e.root_requisition_y();
-                    e.root_allocate_y(ry);
-                    let t2 = time::precise_time_ns();
-                    println!("Layout time: {}", (t2-t1) as f64 * 1.0e-9);
-                },
-                &None => {}
-            }
+    fn layout(&mut self, cairo_ctx: &Context) {
+        if self.layout_required || true {
+            let e = self.root_element.as_root_element().unwrap();
+            let t1 = time::precise_time_ns();
+            let layout_ctx = ElementLayoutContext::new(&self.elem_ctx, cairo_ctx);
+            let rx = e.root_requisition_x(&layout_ctx);
+            e.root_allocate_x(self.width as f64);
+            let ry = e.root_requisition_y();
+            e.root_allocate_y(ry);
+            let t2 = time::precise_time_ns();
+            println!("Layout time: {}", (t2-t1) as f64 * 1.0e-9);
```

Comment:
The problem is that we'll forget about them, which is always the fate of debugging code, which is why so many GUI programs spew vast quantities of no-longer-useful stuff to stdout. Can we enable these prints only in debug mode (with a macro or ...?)?